### PR TITLE
Build and Dockerize a Full-stack React app with Node.js, MySQL and Nginx for reverse proxy

### LINKED ISSUE #2716: Build and Dockerize a Full-stack React application with docker-compose
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

### Introduction paragraph (2-3 paragraphs):

React is one of the common and famous front web frameworks. Nevertheless is most situations, a web application will require back-end services to process different transactions. In this case, React as a front-end framework is not complete, and thus you need a back-end service to make this application complete. On the other hand, docker is the perfect containerizing technology to set up all the environments you need to set up this full-stack application. Docker will help you set React and any front-end environments, then run and build the whole application on a virtual environment using docker.

This tutorial will build a very handy React full-stack application and then dockerize and run the application environments in docker containers. We will dockerize this full-stack React application with Node.js as the back-end service, MySQL as the application database, and NginX for reverse proxy.

### Key takeaways: 

Setup project basics
Build the back-end Node.js application
Build the front-end React application
Create the development environments Docker files
Create the docker-compose.yml file
Run and test the containerized application 

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by mosesreigns**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [ ] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
Fixes #2716

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Good afternoon and thank you for submitting your article for publishing.
Your article has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. :rocket:
Thank you. @mosesreigns

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+### Prerequisites
+
+- Have [Node.js](https://nodejs.org/en/download/) and [Docker](https://www.docker.com/products/docker-desktop) installed on your computer.
+- Prior knowledge on how to use Docker and Docker-compose to run and manage Docker containers
+- Good understanding of using [React](https://www.youtube.com/watch?v=w7ejDZ8SWv8) to create front-end templates
+- Understand How to [create API services using the Node.js](https://www.youtube.com/watch?v=Oe421EPjeBE) framework.
+
+### Setup project basics
+
+This involves two major setups a client and back-end API services. The application involves multi-stage building. You utilize several FROM statements in your Dockerfile for multi-stage builds. Then write a docker-compose.yml file to execute all the Dockerfiles with the proper instructions to set a fully dockerized application. To start with, create a project directory to work on. We will start by creating a Node.js back-end API with Express to expose a REST API and communicate with the MySQL database.
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:23]:
```suggestion
This involves two major setups a client and back-end API services. The application involves multi-stage building. You utilize several FROM statements in your Dockerfile for multi-stage builds. Then write a `docker-compose.yml` file to execute all the Dockerfiles with the proper instructions to set a fully dockerized application. To start with, create a project directory to work on. We will start by creating a Node.js back-end API with Express to expose a REST API and communicate with the MySQL database.
```

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+
+### Build the back-end Node.js application
+
+Node.js will help us to process all tractions and communication that we need to make to the database. In this case, we will use a CRUD use case to implement a simple Node.js Rest API.
+
+#### Step1: Project structure
+
+Inside your project directory, go ahead and create a new folder and name it a server. Will write our backed login inside this directory. So once created cd to this folder and initialize a Node.js application with `npm init -y`. Once executed, a `package.json` file will be created to add any dependencies that we need our application to use.
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:31]:
```suggestion
Inside your project directory, go ahead and create a new folder and name it a `server`. Will write our backed login inside this directory. So once created `cd` to this folder and initialize a Node.js application with `npm init -y`. Once executed, a `package.json` file will be created to add any dependencies that we need our application to use.
```

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+);
+```
+
+#### Step9: Test
+
+Inside the client folder, run `npm start` to check if this is working. Then, open `http://localhost:3000/` and a page similar to this show be sown on your web browser.
+
+![react-application](/engineering-education/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/react-application.png)
+
+### Create the Dockerfiles environments
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:402]: You could reference other Section articles on using Dockers and containers

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+
+```bash
+docker run -it -p 4002:3001 server
+```
+
+Here will are exposing port 3001 (the port that is running the Node.js server application) to port 4002 outside the container—run the above command. And if now you open `http://localhost:4002/` on a browser, you can see that the Node.js home route is okay and the app is working inside Docker.
+
+![dockerized-nodejs-server](dockerized-nodejs-server.png)
+
+### Setting the Nginx server
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:470]: What is proxy? What is reverse proxy? Why do we use Nginx? What other alterative solutions do we have?

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+Now you access the Adminer using route `http://localhost:8000/`.
+
+![adminer](/engineering-education/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/adminer.png)
+
+To log in, use `mysql_db` as the server Username as `root` and password as `MYSQL_ROOT_PASSWORD`. You can see we have database `books` set and table `books_reviews` created.
+
+![adminer-db-and-table](/engineering-education/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/adminer-db-and-table.png)
+
+To start interacting with the application, open `http://localhost:3050/` on a browser. Now you can start adding new records, editing book reviews, and deleting records. In addition, you can refresh the `books_reviews` table on Adminer to view if the changes were implemented in the database.
+
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:668]: You can add a link to the GitHub repository containing the full code

File: content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+- Express - a Node.js framework that enables us to create API endpoints to request data from the server, a process that requests and sends a response with the requested information. Express will help us write different functions that will help us process different CRUD requests.
+- Mysql2 - will enable us to write different functions that will help us access MySQL Server and execute SQL queries from different requests.
+
+Run the following command (inside the server folder) to install the above packages
+
+```bash
+npm install cors express mysql2
+```
+
+#### Step3: Writing the API configurations
```
Comment (srishilesh) [inline content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md:47]: You can probably give idea on what is the objective of this application. What are you trying to build?

**Review by srishilesh — CHANGES_REQUESTED** [review]
@mosesreigns 

Great article.

I've requested for a few changes. Please ping me when done.

Thanks.

**Comment by mosesreigns** [comment]
> @mosesreigns
> 
> Great article.
> 
> I've requested for a few changes. Please ping me when done.
> 
> Thanks.

Thank you @srishilesh. I have made some changes to improved on what you suggested,

**Comment by mosesreigns** [comment]
Any update @srishilesh

**Comment by srishilesh** [comment]
@mosesreigns 

Will review it by tonight.

**Comment by hectorkambow** [comment]
@mosesreigns 
Overall nice job - very informative and detailed tutorial (there were some grammatical issues throughout the articles that needed to be addressed) - please be sure to go over any edits made by the peer reviewer or myself. Articles should come in polished and ready to publish.

cc @srishilesh please do the same- review to improve future reviews - there were some areas that needed to be fixed before we could publish. 
👍

**Pull request merged**